pmd-scala/src/main/java/net/sourceforge/pmd/cpd/ScalaTokenizer.java

```diff
@@ -0,0 +1,119 @@
+/**
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.cpd;
+
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Properties;
+
+import org.apache.commons.lang3.StringUtils;
+
+import net.sourceforge.pmd.lang.LanguageRegistry;
+import net.sourceforge.pmd.lang.LanguageVersion;
+import net.sourceforge.pmd.lang.scala.ScalaLanguageHandler;
+import net.sourceforge.pmd.lang.scala.ScalaLanguageModule;
+
+import scala.collection.JavaConverters;
+import scala.meta.Dialect;
+import scala.meta.inputs.Input;
+import scala.meta.internal.tokenizers.ScalametaTokenizer;
+
+/**
+ * Scala Tokenizer class. Uses the Scala Meta Tokenizer.
+ */
+public class ScalaTokenizer implements Tokenizer {
+
+    /**
+     * Denotes the version of the scala dialect to use. Based on the values in
+     * {@linkplain ScalaLanguageModule#getVersions()}
+     */
+    public static final String SCALA_VERSION_PROPERTY = "scala_version";
+    private final Dialect dialect;
+
+    /**
+     * Create the Tokenizer using properties from the system environment.
+     */
+    public ScalaTokenizer() {
+        this(System.getProperties());
+    }
+
+    /**
+     * Create the Tokenizer given a set of properties.
+     * 
+     * @param properties
+     *            the {@linkplain Properties} object to use
+     */
+    public ScalaTokenizer(Properties properties) {
+        String scalaVersion = properties.getProperty(SCALA_VERSION_PROPERTY);
+        LanguageVersion langVer;
+        if (scalaVersion == null) {
+            langVer = LanguageRegistry.getLanguage(ScalaLanguageModule.NAME).getDefaultVersion();
+        } else {
+            langVer = LanguageRegistry.getLanguage(ScalaLanguageModule.NAME).getVersion(scalaVersion);
+        }
+        dialect = ((ScalaLanguageHandler) langVer.getLanguageVersionHandler()).getDialect();
+    }
+
+    @Override
+    public void tokenize(SourceCode sourceCode, Tokens tokenEntries) throws IOException {
+        String filename = sourceCode.getFileName();
+        // create the full code file
+        String fullCode = StringUtils.join(sourceCode.getCode(), "\n");
+
+        // create the input file for scala
+        Input.VirtualFile vf = new Input.VirtualFile(filename, fullCode);
+        ScalametaTokenizer tokenizer = new ScalametaTokenizer(vf, dialect);
+
+        // tokenize with a filter
+        scala.meta.tokens.Tokens tokens = tokenizer.tokenize();
```

Comment:
the scalameta tokenizer always inserts an EOF for us. I can't seem to get it to throw an exception given any characters, Unicode included. It has a few "catchall" conditions like "identifier" that prevent the normal cases like "inr x=4" instead of "int x=4"  (here "inr" is marked as an "identifier")

Reply:
@tophersmith `inr` is a correct identifier in scala (about anything is IIRC, so I don't think we'll get it to fail that way ^^). Maybe you can try an unclosed string literal, or some XML fragment that's not well-formed

Reply:
Good call on the string literal. That got me a TokenizeException finally. I've added Andreas's suggestion above.